fix: #457 Parameters from `PathItem` can now be overriden in `Operation`

mtovts opened the pull request:
This PR aims to fix the code generation with `PathItems` parameters overriding
in `Operation`. Please, read more in #457 .

Seems like overriding is job for `set` and `__hash__` overriding. But `Property` class is
defined with `@attr.s(auto_attribs=True, frozen=True)`. And how to achieve `__hash__`
overriding with + `frozen=True` is not obvious. It is the reason for using `dict`.

- Parameter location `list`-s replaced with `dict`-s because it allows to
  build `Endpoint.from_data()` with `Operation` parameters then add `PathItem`
  parameters in `EndpointCollection.from_data()` by `dict.setdefault()` to avoid `Operation`
  parameters overriding with `PathItem` parameters.

- Removes from `Endpoint._add_parameters`

    ```python3
    return ParseError(data=param, detail="Parameter must be declared in path or query"), schemas
    ```

  Because parameter with location other than listed in `ParameterLocation(str, Enum)` will
  not be validated by `pydantic`. So why do we need to check that case here twice?

- Add more obvious warning in duplicated parameters case.
- Always use `{param.name}_{param.param_in}` python identifier for parameters because it leads
  to more obvious usage and simplifies the code.
  

  Please, pay attention to `test__add_parameters_happy`
  in `tests\test_parser\test_openapi.py`. I'm not sure if I have done it best way there.

Custom templates in end-to-end tests did not pass. I believe I haven't touched them, but it
fails. I made it passing by making imports relative. I ran it on Windows, so maybe it fixes #447 ?

mtovts commented:
> Always use {param.name}_{param.param_in} python identifier for parameters because it leads
to more obvious usage and simplifies the code.

I shouldn’t have changed that behavior. I brought it back and fixed it for this PR.

dbanty commented on openapi_python_client/templates/endpoint_macros.py.jinja:
```diff
@@ -135,13 +135,13 @@ multipart_data=multipart_data,
 {{ parameter.python_name }}={{ parameter.python_name }},
 {% endfor %}
-{% for parameter in endpoint.cookie_parameters %}
+{% for _, parameter in endpoint.cookie_parameters.items() %}
```
```suggestion
{% for parameter in endpoint.cookie_parameters.values() %}
```

dbanty commented on openapi_python_client/templates/endpoint_module.py.jinja:
```diff
@@ -19,7 +19,7 @@ def _get_kwargs(
     url = "{}{{ endpoint.path }}".format(
         client.base_url
-        {%- for parameter in endpoint.path_parameters -%}
+        {%- for _, parameter in endpoint.path_parameters.items() -%}
```
```suggestion
        {%- for parameter in endpoint.path_parameters.values() -%}
```

dbanty commented on tests/test_parser/test_openapi.py:
```diff
@@ -527,57 +528,79 @@ def test__add_parameters_parse_error(self, mocker):
-            endpoint=endpoint, data=oai.Operation.construct(parameters=[param]), schemas=schemas, config=config
-        )
-        assert result == (ParseError(data=param, detail="Parameter must be declared in path or query"), parsed_schemas)
+        try:
+            oai.Parameter.construct(
+                name="test", required=True, param_schema=mocker.MagicMock(), param_in="error_location"
+            )
+        except Exception as e:
+            assert e is pydantic.ValidationError
```
This test isn't actually checking that the exception happens, only that it's correct if there is one (which there isn't because `construct()` bypasses validation). You'll want something like this:
```suggestion
        with pytest.raises(pydantic.ValidationError):
            oai.Parameter(
                name="test", required=True, param_schema=mocker.MagicMock(), param_in="error_location"
            )
```
Brings up a good point though, we should probably add tests for all the validation code.

[pytest raises docs](https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions)

mtovts replied:
Oh sure. Thank you!

dbanty commented on openapi_python_client/parser/openapi.py:
```diff
@@ -258,37 +279,21 @@ def _add_parameters(
+                same_location: bool = location == param.param_in
 
-            endpoint.relative_imports.update(prop.get_imports(prefix="..."))
+                if existing_prop and not same_location:
```
I think there's a potential for naming conflicts causing issues when the location _is_ the same. For example, if you had "some_name" in both path and query, it would be renamed to "some_name_path" and "some_name_query". If you then had a param named "some_name_path" in path, I think you'd get invalid Python code.

This should either error out here or attempt to further resolve naming issues. I think the former is prudent since we could spend all day trying to de-conflict things and it probably wouldn't be consistent.

Perhaps this was the idea behind `used_python_identifiers`? So new names can be checked?

mtovts replied:
Oops! Sure, I really forgot to add the `used_python_identifiers` check. Sorry. It is added now. Also I added a test for your case.

To resolve conflicts, I wanted to always use the `{param.name}_{param.param_in}` for the parameter identifiers. But I think this is a rare use case. So I put it back as it was before. 

I think it's a good idea to make it an option in the config. Something like `parameters_location_postfix: bool`. This will allow avoiding the use of postfixes with non-duplicated identifiers in most cases. And this will allow avoiding naming conflicts in the case you described, if this option is set.

I added the `parameters with same Python identifier` error. Maybe we should automatically retry generate endpoint with option described above if that error happens.

dbanty commented on openapi_python_client/parser/openapi.py:
```diff
@@ -88,15 +90,16 @@ class Endpoint:
     json_body: Optional[Property] = None
     multipart_body: Optional[Property] = None
     errors: List[ParseError] = field(default_factory=list)
+    used_python_identifiers: Set[PythonIdentifier] = field(default_factory=set)
```
I see where this is written, but not where it's read. Is this residual or is there a missing piece?

mtovts replied:
Added usage.

dbanty reviewed (changes requested):
Thanks for fixing this, and thanks for refactoring that parameter handling stuff! There was a lot of cruft in there, refactoring it was deep down on my todo list 😅.

I have a few changes requested, and I'm going to enable CI on this PR so we can see if it complains about anything else.

mtovts commented on openapi_python_client/parser/openapi.py:
```diff
@@ -258,37 +279,21 @@ def _add_parameters(
+                    prop.set_python_name(new_name=f"{param.name}_{param.param_in}", config=config)
+
+            endpoint.relative_imports.update(prop.get_imports(prefix="..."))
+            parameters_by_location[param.param_in].setdefault(prop.name, prop)
```
I want to use `dict.setdefault()` here because first the `Endpoint` is created to which `Operation` parameters are added. Then added `PathItem` parameters to the `EndpointCollection`. `Operation` parameters has priority over `PathItem` parameters. `dict.setdefault()` allows not to override `Operation` parameters if they are added.

mtovts replied:
I refactored it. And finally used this 😅

mtovts commented:
@dbanty Thank you for the review! Added your suggestions and a test for the case you described. Also added forgotten `used_python_identifiers` usage. It is ready now. Should we create a new issue for the behavior I've suggested to resolve identifiers conflict?

dbanty commented on openapi_python_client/parser/openapi.py:
```diff
@@ -258,37 +279,38 @@ def _add_parameters(
             if isinstance(prop, ParseError):
                 return ParseError(detail=f"cannot parse parameter of endpoint {endpoint.name}", data=prop.data), schemas
 
```
```suggestion
if prop.name in parameters_by_location[param.param_in]:
    # This parameter was defined in the operation, so ignore the PathItem definition
    continue
```

dbanty commented on openapi_python_client/parser/openapi.py:
```diff
@@ -258,37 +279,38 @@ def _add_parameters(
-            else:
-                used_python_names[prop.python_name] = (prop, param.param_in)
+            for location, parameters_dict in parameters_by_location.items():
+                existing_prop: Optional[Property] = parameters_dict.get(prop.name)
+                same_location: bool = location == param.param_in
+
+                if existing_prop and not same_location:
```
```suggestion
                if location == param.param_in or prop.name not in parameters_dict:
                    continue
                existing_prop: Property = parameters_dict[prop.name]
```
This saves a bit of processing time when checking the same location and lets you tab out the following logic, reducing cognitive complexity.

dbanty reviewed (changes requested):
Thanks for keeping up with this! I have one fix to an error message and a couple suggested refactors for easier readability (subjectively).

mtovts commented:
@dbanty Dylan, thanks for these improvements! I agree, and also added the missing test 👀

dbanty commented on openapi_python_client/parser/openapi.py:
```diff
@@ -257,38 +279,40 @@ def _add_parameters(
-                    schemas,
-                )
-            name_check.add(prop.python_name)
+            if prop.name not in parameters_by_location[param.param_in]:
+                if prop.python_name in endpoint.used_python_identifiers:
+                    return (
+                        ParseError(
+                            detail=f"Parameters with same Python identifier `{prop.python_name}` detected", data=data
+                        ),
+                        schemas,
+                    )
+                endpoint.used_python_identifiers.add(prop.python_name)
+                parameters_by_location[param.param_in][prop.name] = prop
```
```suggestion
            if prop.python_name in endpoint.used_python_identifiers:
                return (
                    ParseError(
                        detail=f"Parameters with same Python identifier `{prop.python_name}` detected", data=data
                    ),
                    schemas,
                )
            endpoint.used_python_identifiers.add(prop.python_name)
            parameters_by_location[param.param_in][prop.name] = prop
```

dbanty replied:
Don't need this check here now because we moved it up top.

dbanty reviewed (changes requested):
One more tweak, sorry I didn't clarify in the last round.

dbanty commented:
Everything else looks good so I'm gonna try and tweak this myself real quick so you don't have to jump in here again.

dbanty reviewed (approved):
Thanks again for all your work on this!
